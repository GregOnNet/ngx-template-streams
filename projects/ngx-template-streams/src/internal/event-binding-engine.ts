import { some } from 'fp-ts/lib/Option';
import { escapeQuotes, escapeEventPayload } from '../utils/template-helpers';
import { INTERNAL_PREFIX } from './constants';
import { parseEventBindings } from './event-binding-parser';

type TemplateOperation = (template: string) => string;

export function updateEventBindings(template: string, isInlineTemplate = false) {
  return parseEventBindings(template)
    .chain(transformEventBindings(template, isInlineTemplate))
    .getOrElse(template);
}

function transformEventBindings(template: string, isInlineTemplate: boolean) {
  return (eventBindings: Array<RegExpExecArray>) => {
    return getTemplateOperations(eventBindings, isInlineTemplate).map(executeOperations(template));
  };
}

function getTemplateOperations(eventBindings: Array<RegExpExecArray>, isInlineTemplate: boolean) {
  const operations = eventBindings.map(([fullMatch, event, streamName, eventPayload]) => {
    const eventBinding = createEventBinding(event, streamName, eventPayload, isInlineTemplate);
    return createEventBindingOperation(fullMatch, eventBinding);
  });

  return some(operations);
}

function executeOperations(template: string) {
  return (operations: Array<TemplateOperation>) => {
    return operations.reduce<string>((acc: string, operation: TemplateOperation) => {
      return operation(acc);
    }, template);
  };
}

function createEventBinding(event: string, streamName: string, eventPayload = '$event', isInlineTemplate: boolean) {
  const eventBinding = `(${event})="${INTERNAL_PREFIX}${streamName}.next(${escapeEventPayload(eventPayload)})"`;
  return isInlineTemplate ? eventBinding : escapeQuotes(eventBinding);
}

function createEventBindingOperation(fullMatch: string, replacement: string): TemplateOperation {
  return (s: string) => {
    return s.replace(fullMatch, replacement);
  };
}
