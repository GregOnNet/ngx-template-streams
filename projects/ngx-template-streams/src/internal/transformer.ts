import { tsquery } from '@phenomnomnominal/tsquery';
import { none, some } from 'fp-ts/lib/Option';
import * as ts from 'typescript';
import { updateEventBindings } from './event-binding-engine';

const INLINE_TEMPLATE_QUERY = `ClassDeclaration:has(Decorator:has(Identifier[name="Component"])) PropertyAssignment:has(Identifier[name="template"])`;

export function templateStreamTransformer(context: ts.TransformationContext) {
  return (sourceFile: ts.SourceFile) => {
    return transformInlineTemplate(sourceFile, context).getOrElse(sourceFile);
  };
}

function transformInlineTemplate(sourceFile: ts.SourceFile, context: ts.TransformationContext) {
  return findInlineTemplate(sourceFile).map(updateInlineTemplate(sourceFile, context));
}

function updateInlineTemplate(sourceFile: ts.SourceFile, context: ts.TransformationContext) {
  return (nodes: Array<ts.Node>) => {
    const visitor: ts.Visitor = (node: ts.Node) => {
      if (nodes.includes(node)) {
        const prop = node as ts.PropertyAssignment;

        const currentTemplate = prop.initializer.getFullText();

        const inlineTemplate = ts.createNoSubstitutionTemplateLiteral(
          extractTemplateFromTemplateLiteral(currentTemplate)
            .map(template => updateEventBindings(template, true))
            .getOrElse(currentTemplate)
        );

        return ts.updatePropertyAssignment(prop, prop.name, inlineTemplate);
      }

      return ts.visitEachChild(node, visitor, context);
    };

    return ts.visitNode(sourceFile, visitor);
  };
}

function extractTemplateFromTemplateLiteral(templateLiteral: string) {
  const [, template] = /[`\s'"]*(.*[^\s`'"])[`\s'"]*/s.exec(templateLiteral);
  return some(template);
}

function findInlineTemplate(sourceFile: ts.SourceFile) {
  const nodes = tsquery(sourceFile, INLINE_TEMPLATE_QUERY) as ts.PropertyAssignment[];
  return nodes.length ? some(nodes) : none;
}
