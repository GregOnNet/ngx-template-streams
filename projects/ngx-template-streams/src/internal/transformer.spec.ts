import { readFileSync } from 'fs';
import { resolve } from 'path';
import * as rimraf from 'rimraf';
import { getFixtureFactory } from '../testing/test-helpers';
import compile from '../testing/ts-compiler';
import * as engine from './event-binding-engine';
import * as transformer from './transformer';

const outputDir = resolve(__dirname, '../__test__');
const fixtures = resolve(__dirname, '../fixtures');

const getComponentFixture = getFixtureFactory('component');

describe('Transformer', () => {
  let transformerSpy: jest.SpyInstance;
  let updateEventBindingsSpy: jest.SpyInstance;

  beforeEach(() => {
    transformerSpy = jest.spyOn(transformer, 'templateStreamTransformer');
    updateEventBindingsSpy = jest.spyOn(engine, 'updateEventBindings');
  });

  afterEach(() => {
    rimraf.sync(outputDir);
    jest.restoreAllMocks();
  });

  it('should apply transformation to inline template', () => {
    const { sourceFile, outputFile } = getComponentFixture('inline-template');

    compile(resolve(fixtures, sourceFile), [transformer.templateStreamTransformer]);

    const output = readFileSync(resolve(outputDir, outputFile), 'utf-8');

    expect(output).toMatchSnapshot();
    expect(transformerSpy).toHaveBeenCalled();
    expect(updateEventBindingsSpy).toHaveBeenCalled();
  });

  it('should not apply transformation to external template', () => {
    const { sourceFile, outputFile } = getComponentFixture('external-template');

    compile(resolve(fixtures, sourceFile), [transformer.templateStreamTransformer]);

    const output = readFileSync(resolve(outputDir, outputFile), 'utf-8');

    expect(output).toMatchSnapshot();
    expect(updateEventBindingsSpy).not.toHaveBeenCalled();
  });
});
