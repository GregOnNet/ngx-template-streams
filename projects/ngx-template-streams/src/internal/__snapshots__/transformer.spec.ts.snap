// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transformer should apply transformation to inline template 1`] = `
"import * as tslib_1 from \\"tslib\\";
import { Component } from '@angular/core';
let TestComponent = class TestComponent {
};
TestComponent = tslib_1.__decorate([
    Component({
        selector: 'test',
        template: \`<button (click)=\\"__click$.next('the payload')\\">Click Me</button>\`,
        styleUrls: ['./app.component.css']
    })
], TestComponent);
export { TestComponent };
"
`;

exports[`Transformer should not apply transformation to external template 1`] = `
"import * as tslib_1 from \\"tslib\\";
import { Component } from '@angular/core';
let TestComponent = class TestComponent {
};
TestComponent = tslib_1.__decorate([
    Component({
        selector: 'test',
        templateUrl: './some/path',
        styleUrls: ['./app.component.css']
    })
], TestComponent);
export { TestComponent };
"
`;
